zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -644,73 +652,83 @@ where
         .map_err(Error::DataSource)?
         .ok_or(Error::KeyNotRecognized)?;
 
-    let (sapling_anchor, sapling_inputs) = proposal_step.shielded_inputs().map_or_else(
-        || Ok((sapling::Anchor::empty_tree(), vec![])),
-        |inputs| {
-            wallet_db.with_sapling_tree_mut::<_, _, Error<_, _, _, _>>(|sapling_tree| {
-                let anchor = sapling_tree
-                    .root_at_checkpoint_id(&inputs.anchor_height())?
-                    .into();
+    let (sapling_anchor, sapling_inputs) =
+        if proposal_step.involves(PoolType::Shielded(ShieldedProtocol::Sapling)) {
```

Comment:
I don't fully understand what this extra conditional has to do with allowing the proposer to specify a fallback change pool. But I think it's an optimisation to not construct a protocol's builder if we know we won't be generating any outputs (real or padded) for it, which we now know deterministically after providing a fallback change pool?

Reply:
It's an optimization to avoid constructing the ShardTree.

Reply:
Mmk. Still not clear what necessitated it to happen in the fallback change pool commit, but the optimisation makes sense.

Reply:
Since the Orchard code is being added, this was to make the two implementations consistent with one another.